app/js/lib/server/webpush.js

```diff
@@ -0,0 +1,94 @@
+'use strict';
+
+import EventDispatcher from '../common/event-dispatcher';
+
+// Private members
+const p = Object.freeze({
+  // Properties,
+  api: Symbol('api'),
+  settings: Symbol('settings'),
+
+  // Methods:
+  listenForMessages: Symbol('listenForMessages'),
+});
+
+export default class WebPush extends EventDispatcher {
+  constructor(api, settings) {
+    super(['message']);
+
+    this[p.api] = api;
+    this[p.settings] = settings;
+
+    Object.seal(this);
+  }
+
+  subscribeToNotifications(resubscribe = false) {
+    if (!navigator.serviceWorker) {
+      return Promise.reject('No service worker supported');
+    }
+
+    navigator.serviceWorker.addEventListener('message',
+      this[p.listenForMessages].bind(this));
+
+    const settings = this[p.settings];
+    if (settings.pushEndpoint && settings.pushPubKey && settings.pushAuth &&
+        !resubscribe) {
+      return Promise.resolve();
+    }
+
+    return navigator.serviceWorker.ready
+      .then((reg) => reg.pushManager.subscribe({ userVisibleOnly: true }))
+      .then((subscription) => {
+        const endpoint = subscription.endpoint;
+        const key = subscription.getKey ? subscription.getKey('p256dh') : '';
+        const auth = subscription.getKey ? subscription.getKey('auth') : '';
```

Comment:
Why would it be OK to default to the key `''` if `subscription.getKey` is falsey?  Given that getKey will explicitly return `null` if the named key isn't found in the `PushSubscription`, which subsequently will throw a `TypeError` when attempting to construct a new `Uint8Array` from `null`, should we just throw an `Error` if `getKey` is falsey?  (Although I'm not certain why we check this anyway - implementation inconsistency?)